src/main/java/duke/TaskList.java

```diff
@@ -0,0 +1,158 @@
+     * @param storage Storage to write the changes after removing.
+     * @throws IOException if an error occurs while writing to the output file.
+     */
+    public void remove(int i, Storage storage) throws IOException {
+        Task task = this.getTask(i);
+        String commandString = this.addCommands.get(i);
+        this.tasks.remove(i);
+        this.addCommands.remove(i);
+
+        StringJoiner sj = new StringJoiner("\n", "", "\n");
+        this.addCommands.stream().forEach(c -> sj.add(c));
+        try {
+            storage.write(sj.toString());
+        } catch (IOException e) {
+            this.tasks.add(i, task);
+            this.addCommands.add(i, commandString);
+            throw new IOException("There was a problem writing the change to the file. duke.task.Task not removed.");
+        }
+    }
+
+    /**
+     * Marks a task from the task list, and writes the change to the output file.
+     *
+     * @param i 0-based index of task to be marked.
+     * @param storage Storage to write the changes after marked.
+     * @throws IOException if an error occurs while writing to the output file.
+     */
+    public void mark(int i, Storage storage) throws IOException {
+        this.getTask(i).mark();
+        String commandString = this.addCommands.get(i);
+        if (!commandString.contains("/done")) {
+            this.addCommands.set(i, commandString + " /done");
+        }
+
+        StringJoiner sj = new StringJoiner("\n", "", "\n");
+        this.addCommands.stream().forEach(c -> sj.add(c));
+        try {
+            storage.write(sj.toString());
+        } catch (IOException e) {
+            this.getTask(i).unmark();
+            this.addCommands.set(i, commandString);
+            throw new IOException("There was a problem writing the change to the file. duke.task.Task not marked.");
+        }
+    }
+
+    /**
+     * Unmarks a task from the task list, and writes the change to the output file.
+     *
+     * @param i 0-based index of task to be unmarked.
+     * @param storage Storage to write the changes after unmarked.
+     * @throws IOException if an error occurs while writing to the output file.
+     */
+    public void unmark(int i, Storage storage) throws IOException {
+        this.getTask(i).unmark();
+        String commandString = this.addCommands.get(i);
+        if (commandString.contains("/done")) {
+            this.addCommands.set(i, commandString.replace("/done", ""));
+        }
+
+        StringJoiner sj = new StringJoiner("\n", "", "\n");
+        this.addCommands.stream().forEach(c -> sj.add(c));
+        try {
+            storage.write(sj.toString());
+        } catch (IOException e) {
+            this.getTask(i).mark();
+            this.addCommands.set(i, commandString);
+            throw new IOException("There was a problem writing the change to the file. duke.task.Task not unmarked.");
+        }
+    }
+
+    public TaskList search(String keyword) {
```

Comment:
Remember to add javadoc comments for the methods that you add! 😄

Reply:
Resolved in a future branch!